File: ml/src/wikidetect/detect.py
"""Score one text and phrase the verdict in benchmark terms.

Port of the retired detect.js + scorer.js verdictFor: the strictest
benchmark operating point a score clears gives the measured trade-off at
that level — honest phrasing, not invented confidence labels.
"""

import json

from . import config
from .wiki import split_paragraphs, word_count

_scorers: dict = {}


def get_scorer(model: str = config.DEFAULT_MODEL):
    if model not in _scorers:
        from .models import Scorer

        _scorers[model] = Scorer(model)
    return _scorers[model]


def load_calibration(model: str = config.DEFAULT_MODEL):
    try:
        return json.loads(config.THRESHOLDS.read_text()).get(model)
    except FileNotFoundError:
        return None


def verdict_for(p: float, calibration: dict | None) -> dict:
    if not calibration:
        return {"verdict": "AI-leaning (uncalibrated)" if p >= 0.5 else "human-leaning (uncalibrated)"}
    cleared = next((o for o in reversed(calibration["operatingPoints"]) if p >= o["t"]), None)
    if not cleared:
        return {"verdict": "human-typical"}
    verdict = "AI-typical" if cleared["fpr"] <= 0.05 else "leaning AI — inspect manually"
    return {"verdict": verdict, "atThisLevel": {"caughtAI": cleared["tpr"], "falseFlaggedHuman": cleared["fpr"]}}


def detect_text(text: str, model: str = config.DEFAULT_MODEL) -> dict:
    """The full structured finding the Node side renders."""
    calibration = load_calibration(model)
    scorer = get_scorer(model)

    paragraphs = split_paragraphs(text)
    para_texts = paragraphs if len(paragraphs) > 1 else []
    doc_p, *para_ps = scorer.score_batch([text, *para_texts])

    doc = verdict_for(doc_p, calibration)
    return {
        "model": model,
        "pAI": round(doc_p, 4),
        **doc,
        "shortText": word_count(text) < config.MIN_DOC_WORDS,
        "benchmark": (
            {"auroc": calibration["auroc"], "samples": calibration["samples"]}
            if calibration
            else "uncalibrated — run: wikidetect calibrate"
        ),
        "paragraphs": [
            {
                "pAI": round(p, 4),
                "unreliable": word_count(t) < config.MIN_DOC_WORDS,
                "preview": t[:77] + "..." if len(t) > 80 else t,
            }
            for t, p in zip(para_texts, para_ps)
        ],
    }


def render_human(result: dict) -> str:
    lines = []
    bench = result["benchmark"]
    suffix = (
        f"  (AUROC {bench['auroc']} on {bench['samples']['ai']} AI / {bench['samples']['human']} human Wikipedia samples)"
        if isinstance(bench, dict)
        else ""
    )
    lines.append(f"model      : {result['model']}{suffix}")
    lines.append(f"p(AI)      : {result['pAI']}{'   ⚠ short text — low confidence' if result['shortText'] else ''}")
    lines.append(f"verdict    : {result['verdict']}")
    if "atThisLevel" in result:
        a = result["atThisLevel"]
        lines.append(f"             on the benchmark, scores this high caught {round(a['caughtAI'] * 100)}% of AI text")
        lines.append(f"             while false-flagging {round(a['falseFlaggedHuman'] * 100)}% of human text")
    if result["paragraphs"]:
        lines.append("\nper-paragraph (highest p(AI) first):")
        for p in sorted(result["paragraphs"], key=lambda x: -x["pAI"]):
            mark = "*" if p["unreliable"] else " "
            lines.append(f"  {p['pAI']:.3f}{mark} {p['preview']}")
        if any(p["unreliable"] for p in result["paragraphs"]):
            lines.append(f"  (* under {config.MIN_DOC_WORDS} words — score unreliable)")
    lines.append("\nNote: no detector output is proof of AI authorship. Use this to")
    lines.append("prioritize review; judge content against sources and history.")
    return "\n".join(lines)

File: detector/detect.js
#!/usr/bin/env node
// AI-text triage for Wikipedia cleanup, backed by the best detector on
// this repo's labeled benchmark (currently desklib/ai-text-detector-v1.01;
// see detector/classify.py to re-run the shoot-out).
//
//   node detector/detect.js suspect.txt
//   pbpaste | node detector/detect.js
//   node detector/detect.js --json suspect.txt
//
// Output is p(AI) plus what that number meant on the benchmark of real
// tagged Wikipedia AI text vs pre-ChatGPT human revisions — measured
// catch/false-flag rates, not invented confidence labels.

import { readFileSync } from 'node:fs';
import { startScorer, loadCalibration, verdictFor } from './scorer.js';

const MODEL = 'desklib';
const MIN_WORDS = 40; // detector output on shorter fragments is noise

const argv = process.argv.slice(2);
const asJson = argv.includes('--json');
const file = argv.find((a) => !a.startsWith('--'));
const text = (file ? readFileSync(file, 'utf8') : readFileSync(0, 'utf8')).trim();
if (!text) {
  console.error('usage: node detector/detect.js [--json] [file]  (or pipe text on stdin)');
  process.exit(1);
}

const calibration = loadCalibration(MODEL);
const scorer = startScorer(MODEL);

const paragraphs = text.split(/\n\s*\n/).map((p) => p.trim()).filter(Boolean);
const paraTexts = paragraphs.length > 1 ? paragraphs : [];
const [docP, ...paraPs] = await Promise.all([text, ...paraTexts].map((t) => scorer.score(t)));
scorer.close();

const doc = verdictFor(docP, calibration);
const shortText = text.split(/\s+/).length < MIN_WORDS;

const result = {
  model: MODEL,
  pAI: docP,
  ...doc,
  shortText,
  benchmark: calibration
    ? { auroc: calibration.auroc, samples: calibration.samples }
    : 'uncalibrated — run: detector/venv/bin/python detector/classify.py --model=desklib',
  paragraphs: paraTexts.map((p, i) => ({
    pAI: paraPs[i],
    unreliable: p.split(/\s+/).length < MIN_WORDS,
    preview: p.length > 80 ? p.slice(0, 77) + '...' : p,
  })),
};

if (asJson) {
  console.log(JSON.stringify(result, null, 2));
} else {
  console.log(`model      : ${MODEL}${calibration ? `  (AUROC ${calibration.auroc} on ${calibration.samples.ai} AI / ${calibration.samples.human} human Wikipedia samples)` : ''}`);
  console.log(`p(AI)      : ${docP}${shortText ? '   ⚠ short text — low confidence' : ''}`);
  console.log(`verdict    : ${doc.verdict}`);
  if (doc.atThisLevel) {
    console.log(`             on the benchmark, scores this high caught ${Math.round(doc.atThisLevel.caughtAI * 100)}% of AI text`);
    console.log(`             while false-flagging ${Math.round(doc.atThisLevel.falseFlaggedHuman * 100)}% of human text`);
  }
  if (result.paragraphs.length) {
    console.log('\nper-paragraph (highest p(AI) first):');
    for (const p of [...result.paragraphs].sort((a, b) => b.pAI - a.pAI)) {
      console.log(`  ${p.pAI.toFixed(3)}${p.unreliable ? '*' : ' '} ${p.preview}`);
    }
    if (result.paragraphs.some((p) => p.unreliable)) {
      console.log(`  (* under ${MIN_WORDS} words — score unreliable)`);
    }
  }
  console.log('\nNote: no detector output is proof of AI authorship. Use this to');
  console.log('prioritize review; judge content against sources and history.');
}
